plugins/openssl/openssl.cpp

```diff
@@ -0,0 +1,950 @@
+            ciphertext.ensureCapacity(ciphertextLen);
+
+            // Create and initialize the context
+            encryptCtx = EVP_CIPHER_CTX_new();
+            if (!encryptCtx)
+                failOpenSSLError("creating cipher context");
+
+            // Initialize the envelope
+            std::vector<int> keyLens(keyCount);
+            if (EVP_SealInit(encryptCtx, cipher, encryptedKeys, keyLens.data(), static_cast<byte *>(iv.bufferBase()), publicKeys.data(), keyCount) != static_cast<int>(keyCount))
+                failOpenSSLError("EVP_SealInit");
+
+            // Update the envelope (encrypt the plaintext)
+            int len = 0;
+            if (EVP_SealUpdate(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()), &len, reinterpret_cast<const byte *>(_plaintext), len_plaintext) != 1)
+                failOpenSSLError("EVP_SealUpdate");
+            ciphertextLen = len;
+
+            // Finalize the envelope's ciphertext
+            if (EVP_SealFinal(encryptCtx, static_cast<byte *>(ciphertext.bufferBase()) + len, &len) != 1)
+                failOpenSSLError("EVP_SealFinal");
+            ciphertextLen += len;
+
+            int totalKeyLen = 0;
+            for (int i = 0; i < keyCount; i++)
+                totalKeyLen += keyLens[i];
+
+            // We need to prepend the ciphertext with metadata so the blob can be decrypted;
+            // this is potentially nonstandard
+            MemoryBuffer outBuffer;
+            outBuffer.ensureCapacity(ivLen + (sizeof(size_t)*(keyCount+1)) + totalKeyLen + ciphertextLen);
+            // IV comes first; its length can be derived from the cipher
+            outBuffer.append(ivLen, static_cast<byte *>(iv.bufferBase()));
+            // Number of keys (size_t)
+            outBuffer.append(sizeof(keyCount), reinterpret_cast<byte *>(&keyCount));
+            // Keys; each is (size_t) + <content>
+            for (size_t x = 0; x < keyCount; x++)
+            {
+                size_t keyLen = keyLens[x];
+                outBuffer.append(sizeof(keyLen), reinterpret_cast<byte *>(&keyLen));
+                outBuffer.append(keyLen, encryptedKeys[x]);
+            }
+            // And finally the ciphertext
+            outBuffer.append(ciphertextLen, static_cast<byte *>(ciphertext.bufferBase()));
+
+            // Copy to the ECL result buffer
+            __lenResult = outBuffer.length();
+            __result = outBuffer.detachOwn();
+
+            // Cleanup
+            EVP_CIPHER_CTX_free(encryptCtx);
+            for (size_t x = 0; x < publicKeys.size(); x++)
+                rtlFree(encryptedKeys[x]);
+            rtlFree(encryptedKeys);
+        }
+        catch (...)
+        {
+            if (encryptCtx)
+                EVP_CIPHER_CTX_free(encryptCtx);
+            if (encryptedKeys)
+            {
+                for (size_t x = 0; x < publicKeys.size(); x++)
+                {
+                    if (encryptedKeys[x])
```

Comment:
minor: A general point, for free, delete and rtlFree it is legal to pass a null ptr, and it is defined to do nothing - so no need to check first.